src/java/com/unifina/datasource/DataSource.java

```diff
@@ -1,46 +1,62 @@
 package com.unifina.datasource;
 
-import com.unifina.data.*;
-import com.unifina.domain.data.Feed;
-import com.unifina.feed.AbstractFeed;
-import com.unifina.service.FeedService;
+import com.streamr.client.protocol.message_layer.StreamMessage;
+import com.streamr.client.utils.StreamPartition;
+import com.unifina.data.Event;
+import com.unifina.data.EventQueueMetrics;
+import com.unifina.feed.MessageRouter;
+import com.unifina.feed.StreamMessageSource;
+import com.unifina.feed.map.MapMessageEventRecipient;
+import com.unifina.serialization.SerializationRequest;
 import com.unifina.signalpath.AbstractSignalPathModule;
+import com.unifina.signalpath.AbstractStreamSourceModule;
 import com.unifina.signalpath.SignalPath;
+import com.unifina.signalpath.StopRequest;
 import com.unifina.utils.Globals;
-import grails.util.Holders;
 import org.apache.log4j.Logger;
 
 import java.util.*;
+import java.util.function.Consumer;
 
 /**
- * DataSource is a class global to the current run context. It handles
- * the creation of data feeds either explicitly (via DataSource#connectFeed(feed))
- * or implicitly according to a SignalPath's needs (via DataSource#connectSignalPath(signalPath)).
+ * DataSource wires together SignalPaths and streams of messages and other Events.
+ * There are two types: HistoricalDataSource and RealtimeDataSource.
  *
- * @author Henri
+ * The DataSource is started with DataSource#start(), which starts the event loop
+ * and blocks until done.
  */
-public abstract class DataSource {
+public abstract class DataSource implements Consumer<Event> {
 	private static final Logger log = Logger.getLogger(DataSource.class);
 
 	private final Set<SignalPath> signalPaths = new HashSet<>();
 	private final List<IStartListener> startListeners = new ArrayList<>();
 	private final List<IStopListener> stopListeners = new ArrayList<>();
-	private final Map<Long, AbstractFeed> feedById = new HashMap<>();
-	private final Globals globals;
-	private final boolean isHistoricalFeed;
+	protected final Globals globals;
```

Comment:
- `DataSource` and `Globals` have a 1-to-1 relationship
- Multiple `SignalPath`s can run in a `SignalPathRunner` under the same `(DataSource, Globals)` pair (perhaps this is what you're remembering?)

Nowadays we're always running just one `SignalPath` per `SignalPathRunner`. It doesn't add almost any code complexity to support several, but I think I'll remove the possibility anyway since we're not using the feature.